src/planet-a/icosahedron.cpp

```diff
@@ -0,0 +1,220 @@
+
+    double const pnt = 2.0/5.0 * sqrt(5.0);
+    double const hei = 1.0/sqrt(5.0);
+    double const cxA = 1.0/2.0 - sqrt(5.0)/10.0;
+    double const cxB = 1.0/2.0 + sqrt(5)/10.0;
+    double const syA = 1.0/10.0 * sqrt( 10.0 * (5.0+sqrt(5.0)) );
+    double const syB = 1.0/10.0 * sqrt( 10.0 * (5.0-sqrt(5.0)) );
+
+    std::array<osp::Vector3d, 12> const icosahedronVrtx
+    {{
+        { 0.0,    0.0,    1.0}, // top point
+
+        { pnt,   0.0f,    hei}, // 1 top pentagon
+        { cxA,   -syA,    hei}, // 2
+        {-cxB,   -syB,    hei}, // 3
+        {-cxB,    syB,    hei}, // 4
+        { cxA,    syA,    hei}, // 5
+
+        {-pnt,   0.0f,   -hei}, // 6 bottom pentagon
+        {-cxA,   -syA,   -hei}, // 7
+        { cxB,   -syB,   -hei}, // 8
+        { cxB,    syB,   -hei}, // 9
+        {-cxA,    syA,   -hei}, // 10
+
+        {0.0f,   0.0f,  -1.0f}  // 11 bottom point
+    }};
+
+    // Create the skeleton
+
+    SubdivTriangleSkeleton skeleton;
+
+    // Create initial 12 vertices
+
+    for (SkVrtxId &rVrtxId : vrtxIds)
+    {
+        rVrtxId = skeleton.vrtx_create_root();
+    }
+
+    rPositions.resize(skeleton.vrtx_ids().capacity());
+    rNormals.resize(skeleton.vrtx_ids().capacity());
+
+    // Set positions and normals of the new vertices
+
+    double const scale = std::pow(2.0, pow2scale);
+
+    for (int i = 0; i < gc_icoVrtxCount; i ++)
+    {
+        rPositions[i] = osp::Vector3l(icosahedronVrtx[i] * radius * scale);
+        rNormals[i] = osp::Vector3(icosahedronVrtx[i]);
+    }
+
+    // Add initial triangles
+
+    // Create 20 root triangles by forming 5 groups. each group is 4 triangles
+    skeleton.tri_group_reserve(skeleton.tri_ids().size() + 5);
+
+    auto const vrtx_id_lut = [&vrtxIds] (int i) -> std::array<SkVrtxId, 3>
+    {
+        return { vrtxIds[ sc_icoTriLUT[i][0] ],
+                 vrtxIds[ sc_icoTriLUT[i][1] ],
+                 vrtxIds[ sc_icoTriLUT[i][2] ] };
+    };
+
+    for (int i = 0; i < gc_icoTriCount; i += 4)
+    {
+        //std::array<std::array<SkVrtxId, 3>, 4> ;
+        SkTriGroupId const groupId = skeleton.tri_group_create(0, osp::id_null<SkTriId>(),
+        {
+            vrtx_id_lut(i + 0), vrtx_id_lut(i + 1),
+            vrtx_id_lut(i + 2), vrtx_id_lut(i + 3)
+        });
+
+        triIds[i + 0] = tri_id(groupId, 0);
+        triIds[i + 1] = tri_id(groupId, 1);
+        triIds[i + 2] = tri_id(groupId, 2);
+        triIds[i + 3] = tri_id(groupId, 3);
+    }
+
+    return skeleton;
+}
+
+void subdiv_curvature(
+        double radius, float scale, osp::Vector3l const a, osp::Vector3l const b,
+        osp::Vector3l &rPos, osp::Vector3 &rNorm) noexcept
+{
+    osp::Vector3l const avg = (a + b) / 2;
+    osp::Vector3d const avgDouble = osp::Vector3d(avg) / scale;
+    float const avgLen = avgDouble.length();
+    float const roundness = radius - avgLen;
+
+    rNorm = osp::Vector3(avgDouble / avgLen);
+    rPos = avg + osp::Vector3l(rNorm * roundness * scale);
```

Comment:
This is a function that can easily be vectorized if the need arises. I imagine it might be called a lot, and AVX is a fantastic way of minimizing the performance cost of the `sqrt()` inherent in calculating a vector norm. Obviously not necessary to do it now, just thought I'd point it out. Is this function going to be called as frequently as I think it is?

Reply:
This is used to calculate skeleton vertices only along the edges of a chunk. I'd assume the meat you might be looking for is the hundreds more mesh vertices that need to be calculated to fill the chunk's triangular lattice, found [here](https://github.com/TheOpenSpaceProgram/osp-magnum/blob/a2788e9346825536ec01bf559802476b8bc595ea/src/planet-a/Active/SysPlanetA.cpp#L213-L233). That ChunkVrtxSubdivLUT class had calculated ahead of time in the correct order which pairs of vertices need to be subdivided.

I see this as a bit of a challenge as subdivision is very order-dependent. Vertices needed for the current subdivision are results from the previous subdivision. Curious to see what you might come up with.